### PR TITLE
methylation summary module

### PR DESCRIPTION
**Pull request by ewafula**
#### What scientific question is your analysis addressing?
Summarize preprocessed `Illumina Infinium HumanMethylation  BeadArrays (450K)` methylation measurements produced by the [OpenPedCan methylation-analysis module](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/158) in a summary table that will be utilized with OPenPedCan plotting API and displayed on the NCI MTP portal.
   

#### What was your approach?

Utilize the TARGET preprocessed array methylation measurements/metadata and  RNA-Seq expression matrix to create beta and M-values methylation matrices, calculate probe-level beta quantiles and beta-tpm correlations, and create a summary table of the results. 

#### What GitHub issue does your pull request address?
- https://github.com/PediatricOpenTargets/ticket-tracker/issues/269

#### Which areas should receive a particularly close look?
Review the entire code and test run the module. Also, check the results produced by the module  for consistency 

#### Is there anything that you want to discuss further?
 Please read the module README for detailed instructions on how to run the module, including  the required  input datasets

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes.

#### What types of results are included (e.g., table, figure)?
TSV and JSON files

#### What is your summary of the results?
- `results/methylation-probe-annotations.tsv.gz`
- `results/methylation-beta-values-matrix.tsv.gz`
- `results/methylation-m-values-matrix.tsv.gz`
- `results/methylation-probe-beta-quantiles.tsv.gz`
- `results/methylation-probe-beta-tpm-correlations.tsv.gz`
- `results/methylation-beta-values-summary.tsv.gz`
- `results/methylation-beta-values-summary.jsonl.gz`

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

File: analyses/methylation-summary/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+#### `05-create-methylation-summary-table.R`
+This script creates Pediatric OpenTargets methylation summary table that will be utilized with OPenPedCan plotting API and displayed on the NCI MTP portal.
+
+```
+Rscript --vanilla 05-create-methylation-summary-table.R
+```
+
+#### `06-methyly-summary-tsv2jsonl.py`
```
Comment (jharenza) [inline analyses/methylation-summary/README.md]:
```suggestion
#### `06-methyl-summary-tsv2jsonl.py`
```

File: analyses/methylation-summary/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    - **datatypeId**: Illumina_methylation_array
+    - **chop_uuid**: generate UUID
+    - **datasourceId**: chop_gene_level_methylation
+
+
+## General usage of scripts
+Analyses involving 850k arrays with large number of samples representing OPenPedCan cancer groups (as in the CBTN cohort) will require `~256gb` of memory to run successfully. In some instances the `system /tmp` is too small to hold temporary files generated during analysis by R scripts. Users are advised to create create a `./tmp` in the module directory then execute R script by prepending with TMP/TMPDIR environmental variable as illustrated in the wrapper module bash script, `run-methylation-summary.sh`.
+
+
+#### `Rrequired R packages`
```
Comment (jharenza) [inline analyses/methylation-summary/README.md]:
```suggestion
#### `Required R packages`
```

File: analyses/methylation-summary/README.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+```
+- tidyverse
+- rtracklayer
+- data.table
+- ids
+```
+
+#### `Rrequired Python3 modules`
```
Comment (jharenza) [inline analyses/methylation-summary/README.md]:
```suggestion
#### `Required Python3 modules`
```

File: analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+
+
+"""
+06-methyly-summary-tsv2jsonl.py
```
Comment (jharenza) [inline analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py]:
```suggestion
06-methyl-summary-tsv2jsonl.py
```

File: analyses/methylation-summary/01-create-methylation-annotation-table.R [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+if (!dir.exists(results_dir)) {
+  dir.create(results_dir)
+}
+
+# Function to parse preprocessed methylation results and annotate array
+# probes with GENCODE gene symbols and  Ensembl IDs
+annotate_array_probes <- function(meth_file) {
+  # read UCSC Ensembl to RefSeq mapping file
+  ensembl2refseq <- data.table::fread(
+    file.path(input_dir, "UCSC_hg19-GRCh37_Ensembl2RefSeq.tsv"),
```
Comment (jharenza) [inline analyses/methylation-summary/01-create-methylation-annotation-table.R:36]: Where is this file coming from? Was it generated (if so, we should add a script), if obtained, we should add a link.

File: analyses/methylation-summary/05-create-methylation-summary-table.R [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  showProgress = FALSE) %>% 
+  tibble::as_tibble() %>% 
+  dplyr::right_join(summary_table, by = "Probe_ID")
+
+# Add PMTL disgnation corresponding to GENCODE version 38 (Ensembl 104) gene
+# symbols and Ensembl IDs
+message("Including PMTL disgnation for corresponding Ensembl IDs ...\n")
+pmtl_ensembl_ids <- readr::read_tsv(file.path(gene_match_input_dir, 
+                                              "PMTL_v1.1.tsv"),
+                           show_col_types = FALSE) %>%
```
Comment (jharenza) [inline analyses/methylation-summary/05-create-methylation-summary-table.R]:
```suggestion
pmtl_ensembl_ids <- readr::read_tsv(file.path(data_dir, 
                                              "ensg-hugo-pmtl-mapping.tsv"),
                           show_col_types = FALSE) %>% 
```
should change this to use the file in `data/`

File: analyses/methylation-summary/README.md (hunk 2 above)
Comment (jharenza) [reply analyses/methylation-summary/README.md]: can you confirm all are in the dockerfile, and if not, please add them in a separate PR.

File: analyses/methylation-summary/README.md (hunk 3 above)
Comment (jharenza) [reply analyses/methylation-summary/README.md]: can you confirm all are in the dockerfile, and if not, please add them in a separate PR.

File: analyses/methylation-summary/README.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+```
+Rscript --vanilla 05-create-methylation-summary-table.R
+```
+
+#### `06-methyly-summary-tsv2jsonl.py`
+This script create a JSONL file for Pediatric OpenTargets methylation summary table utilized on NCI MTP portal.
+
+```
+python3 06-methyly-summary-tsv2jsonl.py
```
Comment (jharenza) [inline analyses/methylation-summary/README.md]:
```suggestion
python3 06-methyl-summary-tsv2jsonl.py
```

File: analyses/methylation-summary/run-methylation-summary.sh [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+TMP=./tmp TMPDIR=./tmp Rscript --vanilla 03-calculate-beta-quantiles.R
+
+############ Calculate probe-level correlations between Beta and TPM values ###############
+python3 04-beta-tpm-correlation.py
+
+############################ Create methylation summary table #############################
+TMP=./tmp TMPDIR=./tmp Rscript --vanilla 05-create-methylation-summary-table.R
+
+########################### Convert JSON to JSON Lines (JSONL) ############################
+python3 06-methyly-summary-tsv2jsonl.py
```
Comment (jharenza) [inline analyses/methylation-summary/run-methylation-summary.sh]:
```suggestion
python3 06-methyl-summary-tsv2jsonl.py
```

File: analyses/methylation-summary/README.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+- Using `03-calculate-beta-quantiles.R` script and the methylation `Beta-values matrix`, calculate probe-level `quantiles` for each cancer type (cancer_group) within an OpenPedCan cohort (`results/methyl-probe-beta-quantiles.tsv.gz`). 
+- Using `04-beta-tpm-correlation.R` script and the methylation `Beta-values matrix`, calculate probe-level `correlations` between `RNA-Seq TPM-values` and `Beta-values` for each cancer type (cancer_group) within an OpenPedCan cohort for patients who have both datasets (`results/methyl-probe-beta-tpm-correlations.tsv.gz`). Correlations for cohorts will have independent scripts becuase of the differences in how samples with patients in both RNA-Seq and methylation data are determined. 
+- Summarize all results using `05-create-methylation-summary-table.R` and `06-methyly-summary-tsv2jsonl.py` scripts into a methylation summary table (`results/methyl-beta-values-summary.rds`, `results/methyl-beta-values-summary.tsv.gz` and `results/methyl-beta-values-summary.jsonl.gz`) that will be utilized with OPenPedCan plotting API and displayed on the NCI MTP portal with the following columns:
```
Comment (jharenza) [inline analyses/methylation-summary/README.md]:
```suggestion
- Summarize all results using `05-create-methylation-summary-table.R` and `06-methyl-summary-tsv2jsonl.py` scripts into a methylation summary table (`results/methyl-beta-values-summary.rds`, `results/methyl-beta-values-summary.tsv.gz` and `results/methyl-beta-values-summary.jsonl.gz`) that will be utilized with OPenPedCan plotting API and displayed on the NCI MTP portal with the following columns:
```

**Review by jharenza — CHANGES_REQUESTED** [review]
Hi @ewafula thanks for this PR. Just had a few suggestions, non-urgent, that we should address before this PR is merged.

File: analyses/methylation-summary/04-beta-tpm-correlation.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+import re
+import os
+import sys
+import numpy as np
+import pandas as pd
+import pyreadr as rds
+
+
+# establish base dir
+root_dir = os.path.abspath("../../")
```
Comment (jharenza) [inline analyses/methylation-summary/04-beta-tpm-correlation.py]:
```suggestion
root_dir = os.path.abspath("../../")
```

Can this be more explicit such as the finding git path in R?

File: analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import csv
+import gzip
+import json
+import numpy as np
+import pandas as pd
+from collections import OrderedDict
+
+
+# establish base dir
+root_dir = os.path.abspath("../../")
```
Comment (jharenza) [inline analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py]: Same comment as above

File: analyses/methylation-summary/04-beta-tpm-correlation.py (hunk 10 above)
Comment (ewafula) [reply analyses/methylation-summary/04-beta-tpm-correlation.py]: updated using the `gitpython` module

```
root_dir = git.Repo('.', search_parent_directories=True).working_tree_dir
```

File: analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py (hunk 11 above)
Comment (ewafula) [reply analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py]: updated using the `gitpython` module

```
root_dir = git.Repo('.', search_parent_directories=True).working_tree_dir

File: analyses/methylation-summary/README.md (hunk 9 above)
Comment (ewafula) [reply analyses/methylation-summary/README.md]: fixed

File: analyses/methylation-summary/run-methylation-summary.sh (hunk 8 above)
Comment (ewafula) [reply analyses/methylation-summary/run-methylation-summary.sh]: fixed

File: analyses/methylation-summary/README.md (hunk 7 above)
Comment (ewafula) [reply analyses/methylation-summary/README.md]: fixed

File: analyses/methylation-summary/05-create-methylation-summary-table.R (hunk 6 above)
Comment (ewafula) [reply analyses/methylation-summary/05-create-methylation-summary-table.R]: fixed and adjusted  code to filter for PMTL ESNG IDs  and read proper column header.

File: analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py (hunk 4 above)
Comment (ewafula) [reply analyses/methylation-summary/06-methyly-summary-tsv2jsonl.py]: fixed

File: analyses/methylation-summary/README.md (hunk 1 above)
Comment (ewafula) [reply analyses/methylation-summary/README.md]: fixed

File: analyses/methylation-summary/README.md (hunk 3 above)
Comment (ewafula) [reply analyses/methylation-summary/README.md]: I have confirmed all `R package `and `python 3 modules` are in the docker file, including the python non-standard libraries used in this module (`numpy`, `pandas`, `pyreadr`, and `GitPython`)

**Comment by ewafula** [comment]
@jharenza,  all updates in your review above have applied.  Should be ready to merge.

File: analyses/methylation-summary/README.md (hunk 2 above)
Comment (ewafula) [reply analyses/methylation-summary/README.md]: This was resolved similar comment here: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/165#discussion_r929048910

`I have confirmed all R package and python 3 modules are in the docker file, including the python non-standard libraries used in this module (numpy, pandas, pyreadr, and GitPython)`

File: analyses/methylation-summary/01-create-methylation-annotation-table.R (hunk 5 above)
Comment (jharenza) [reply analyses/methylation-summary/01-create-methylation-annotation-table.R:36]: what about this last one @ewafula ?

Comment (ewafula) [reply analyses/methylation-summary/01-create-methylation-annotation-table.R:36]: I have added to the README input files list where the `Ensembl2RefSeq` gene IDs mapping file was downloaded  - UCSC Table Browser (https://genome.ucsc.edu/cgi-bin/hgTables).

**Review by jharenza — APPROVED** [review]

**Pull request merged**